dimagi/commcare-hq
# Web Apps accessibility improvements - make apps and menus keyboard navigable

## Summary
https://dimagi-dev.atlassian.net/browse/USH-394

This is version 2 of [an earlier change](https://github.com/dimagi/commcare-hq/pull/28902) that was reverted. It makes applications, menu items and case list items behave like links, generally using the following steps:

1. add the role="link" attribute which informs screen readers to treat it as a link.
2. put the element into the tab order with the tabindex="0" attribute, so it can receive keyboard focus.
3. add a keyboard event handler to the Marionette view for the element which treats Enter key presses like mouse clicks.
4. add a label with either aria-labelledby or aria-label to tell a screen reader what text to read for the element.

The flaw in version 1 was that the CaseView Marionette view had an ARIA label and role that prevented users from hearing the individual table data elements with a screen reader, so they no longer have those attributes.

This version also makes the delete button (trashcan) for incomplete sessions work with the keyboard and it avoids any changes to pagination controls to prevent a conflict with another concurrent change.

## Product Description
For users without a screen reader, this change enables keyboard navigation of Web Apps, using the Tab and Enter keys. Navigating by keyboard is one way to test that it works. No other changes should be evident without a screen reader.

## Safety Assurance


### QA Plan

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-2791)

This change should also be reviewed in staging by our screen reader expert who flagged the issue with the previous version of the change.

### Safety story
WAI-ARIA attributes are designed to have no effect outside of accessibility tools, so they shouldn't cause any regressions.

The other changes are putting elements into the focus ring and handling key clicks on them, which should not affect existing users who navigate by mouse clicks.

### Rollback instructions
